Implementing SymmetricDifference class, #8179

kaichogami opened the pull request:
Created a class to calculate the Symmetric Difference of two finite sets(see #8179) . This class structure is similar to that of already implemented `Complement` class. `__xor__` method in `FiniteSet` class is overridden to call `SymmetricDifference` class. 
Usage Example

```
In [1]: from sympy import FiniteSet, SymmetricDifference
In [2]: SymmetricDifference(FiniteSet(1,2,3,4), FiniteSet(3,4,5))
Out[2]: {1, 2, 5}
```

kaichogami commented:
@pelegm I added `NotImplementedError` to `Set`, `Interval`,  `UniversalSet`, `EmptySet` classes. `UniversalSet`, `EmptySet` should just return `other`, although I think as it is not in the scope of the issue #8179, we should create a new issue or directly a PR to make the changes in the above four classes.
Also if you think I missed something, let me know!

hargup commented:
>  I added `NotImplementedError` to `Set`, `Interval`,  `UniversalSet`, `EmptySet` classes. `UniversalSet`, `EmptySet` should just return `other`, although I think as it is not in the scope of the issue #8179

Extending the PR to handle other Set classes shouldn't be difficult if you use the `Complement` and `Union` classes to represent Symmetric Difference. See https://en.wikipedia.org/wiki/Symmetric_difference#Properties (this was also mention by @pelegm in https://github.com/sympy/sympy/issues/8179#issuecomment-58139786). You don't need to write the logic to actually evaluate the Symmmetric Difference. The reduce functions in `Complement` and `Union` classes can handle that.

hargup commented on sympy/sets/tests/test_sets.py:
```diff
@@ -831,3 +831,12 @@ def test_Eq():
+   assert SymmetricDifference(FiniteSet(2,3,4), FiniteSet(2,3,4,5)) == FiniteSet(5)
+   assert FiniteSet(1,2,3,4,5) ^ FiniteSet(1,2,5,6) == FiniteSet(3,4,6)
+
+   #Not Implemented yet
```
You can create a test function with the `@xfail` decorator. `xfail` implies `expected to fail`.

kaichogami commented:
@hargup using the classes made it easy. If you think I still missed something, let me know!

asmeurer commented:
Returning an unevaluated object is better than raising NotImplementedError.

skirpichev commented on sympy/sets/sets.py:
```diff
@@ -206,6 +206,12 @@ def _complement(self, other):
 
+    def _symmetricdifference(self, other):
+        """
+        This function should only be used internally
```
I think, you can remove this docstring.  Method is private if it starts with underscore.

skirpichev commented on sympy/sets/sets.py:
```diff
@@ -882,6 +891,13 @@ def _complement(self, other):
         return Set._complement(self, other)
 
 
+    def _symmetricdifference(self, other):
```
This is exactly same method as for Set.  Remove?

skirpichev commented on sympy/sets/sets.py:
```diff
@@ -1716,6 +1738,12 @@ def _union(self, other):
+        """
+        This function should only be used internally
+        """
+        return Union(Complement(self, other), Complement(other, self))
```
again, same as for Set

skirpichev commented on sympy/sets/tests/test_sets.py:
```diff
@@ -831,3 +831,15 @@ def test_Eq():
 
     assert Eq(s1*s2, s1*s2)
     assert Eq(s1*s2, s2*s1) == False
+
```
add newline

skirpichev commented on sympy/sets/sets.py:
```diff
@@ -206,6 +206,12 @@ def _complement(self, other):
         elif isinstance(other, FiniteSet):
             return FiniteSet(*[el for el in other if self.contains(el) != True])
 
+    def _symmetricdifference(self, other):
```
Probably, we should have a pair of private and public methods, just like other cases.  Also, please see documentation of [builtin set](https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset) class.  I think, that's why public method should be named as `symmetric_difference`.

skirpichev commented on sympy/sets/sets.py:
```diff
@@ -1808,6 +1836,43 @@ def __lt__(self, other):
 
 
+class SymmetricDifference(Set):
+    """Representes the Symmetric Difference of a set with another set.
```
Represents

skirpichev replied:
You could add reference here, i.e. on the wiki page.

kaichogami replied:
@skirpichev Could you tell me what you mean by "You could add reference here, i.e. on the wiki page."? I added the definition from the wikipedia page in docstring.

skirpichev replied:
Yes, but you could add citation of this page right here (with `[1]_` in an appropriate place).

skirpichev replied:
See also [this](https://github.com/sympy/sympy/wiki/Writing-documentation) about writing documentation.

skirpichev commented on sympy/sets/sets.py:
```diff
@@ -1808,6 +1836,43 @@ def __lt__(self, other):
+    {1, 2, 4, 5}
+
+    See Also
+    ========
```
add newline after section header, same below

skirpichev commented on sympy/sets/sets.py:
```diff
@@ -1808,6 +1836,43 @@ def __lt__(self, other):
+
+    @staticmethod
+    def reduce(A, B):
+
```
remove this newline

kaichogami commented:
@skirpichev I made the said changes. I have still lot to learn. 
This looks all set to go to me now.

kaichogami commented:
@skirpichev @pelegm @hargup Looks like its done now.

skirpichev commented:
+1

But please fix merge conflicts.  Also, please squash all commits (except the first) and edit commit message: you can do this with "git rebase".

kaichogami commented:
@skirpichev I squashed the commits, except the first one, into a single commit after fixing the merge conflicts. For what reason are squashing the commits?

skirpichev commented:
I still see two commits.

> For what reason are we squashing the commits?

To cleanup the git history.  Also,  "first commit to solve bug #8179" is not a [good commit message](https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages).

skirpichev commented on sympy/sets/sets.py:
```diff
@@ -206,6 +206,12 @@ def _complement(self, other):
+    def symmetric_difference(self, other):
+         return SymmetricDifference(self, other)
+
+    def _symmetricdifference(self, other):
```
probably, this name should match public method?

kaichogami commented:
@skirpichev I squashed all the commits into one and wrote a better commit message. Looks much cleaner now.

kaichogami committed:
Symmetric Difference: Implemented the class, wrote tests
Implemented a class to calculate the symmetric difference between two given sets. Uses the logic of reduce function of Union and Complement class. '^' operator is overided to call this class. This closes #8179.

Example:
In [1]: from sympy import FiniteSet
In [3]: FiniteSet(1,2,3) ^ FiniteSet(2,3,4)
Out[3]: {1, 4}

hargup commented:
+1, will merge in 24h if no objection. @kaichogami you should also add printer for the `SymmetricDifference` class but you can save that for another PR.

kaichogami commented:
@hargup alright. I'll create a new PR as soon as I add the printer.

hargup commented:
Thanks it is in.
